File: packages/react-native-performance/android/src/main/kotlin/com/shopify/reactnativeperformance/RenderCompletionEventEmitter.kt
package com.shopify.reactnativeperformance

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactContext
import com.facebook.react.modules.core.DeviceEventManagerModule

object RenderCompletionEventEmitter {
  private const val RENDER_COMPLETION_EVENT_NAME = "@shopify/react-native-performance/onRenderComplete"

  fun onRenderComplete(
    context: ReactContext,
    destinationScreen: String,
    renderPassName: String,
    interactive: Interactive,
    componentInstanceId: String
  ) {
    val event = Arguments.createMap().apply {
        putString("timestamp", System.currentTimeMillis().toString())
        putString("renderPassName", renderPassName)
        putString("interactive", interactive.toString())
        putString("destinationScreen", destinationScreen)
        putString("componentInstanceId", componentInstanceId)
    }
    context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        .emit(RENDER_COMPLETION_EVENT_NAME, event)
  }
}

File: packages/react-native-performance/ios/ReactNativePerformance/RenderCompletionEventEmitter.swift
import Foundation

@objc(RenderCompletionEventEmitter)
class RenderCompletionEventEmitter: RCTEventEmitter {
    private static let RENDER_COMPLETION_EVENT_NAME = "@shopify/react-native-performance/onRenderComplete"
    private var hasListeners = false
    private(set) static var INSTANCE: RenderCompletionEventEmitter? = nil

    override init() {
        super.init()
        RenderCompletionEventEmitter.INSTANCE = self
    }

    @objc override func supportedEvents() -> [String]! {
        return [RenderCompletionEventEmitter.RENDER_COMPLETION_EVENT_NAME]
    }

    func onRenderComplete(
        destinationScreen: String,
        renderPassName: String,
        interactive: Interactive,
        componentInstanceId: String
    ) {
        if (hasListeners) {
            let timestamp = Timestamp.nowMillis()
            sendEvent(withName: RenderCompletionEventEmitter.RENDER_COMPLETION_EVENT_NAME, body: [
                "timestamp": String(timestamp),
                "renderPassName": renderPassName,
                "interactive": interactive.description,
                "destinationScreen": destinationScreen,
                "componentInstanceId": componentInstanceId
            ])
        }
    }

    @objc override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    @objc override func startObserving() {
        super.startObserving()
        hasListeners = true
    }

    @objc override func stopObserving() {
        super.stopObserving()
        hasListeners = false
    }
}
